base16ct/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+//!
+//! Adapted from this C++ implementation:
+//!
+//! <https://github.com/Sc00bz/ConstTimeEncoding/blob/master/hex.cpp>
+//!
+//! Copyright (c) 2014 Steve "Sc00bz" Thomas (steve at tobtu dot com)
+//! Derived code is dual licensed MIT + Apache 2.0 (with permission from @Sc00bz)
+
+#[cfg(feature = "alloc")]
+#[macro_use]
+extern crate alloc;
+#[cfg(feature = "std")]
+extern crate std;
+
+use core::fmt;
+
+#[cfg(feature = "alloc")]
+use alloc::{string::String, vec::Vec};
+
+/// Error type
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub enum Error {
+    /// Invalid encoding of provided Base16 string.
+    InvalidEncoding,
+
+    /// Insufficient output buffer length.
+    InvalidLength,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidEncoding => f.write_str("invalid Base16 encoding"),
+            Error::InvalidLength => f.write_str("invalid Base16 length"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error {}
+
+/// Base16 encoding (a.k.a. hexadecimal)
+#[derive(Copy, Clone, Debug, Default, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct Base16 {
+    /// Upper or lower case
+    case: Case,
+}
+
+impl Base16 {
+    /// Lower case hex: 0-9 a-f
+    pub fn lower_case() -> Base16 {
+        Base16 { case: Case::Lower }
+    }
+
+    /// Upper case hex: 0-9 A-F
+    pub fn upper_case() -> Base16 {
+        Base16 { case: Case::Upper }
+    }
+}
+
+impl Base16 {
+    /// Decode a Base16 (hex) string into the provided destination buffer.
+    pub fn decode<'a>(&self, src: impl AsRef<[u8]>, dst: &'a mut [u8]) -> Result<&'a [u8], Error> {
+        let src = src.as_ref();
+
+        let dst_length = Self::decoded_len(src)?;
+        if dst_length > dst.len() {
+            return Err(Error::InvalidLength);
+        }
+
+        let mut err: usize = 0;
+
+        for (i, dst_byte) in dst.iter_mut().enumerate().take(dst_length) {
+            let src_offset = i * 2;
+            let byte = (self.case.decode_nibble(src[src_offset]) << 4)
+                | self.case.decode_nibble(src[src_offset + 1]);
+            err |= byte >> 8;
+            *dst_byte = byte as u8;
+        }
+
+        if err == 0 {
+            Ok(&dst[..dst_length])
+        } else {
+            Err(Error::InvalidEncoding)
+        }
+    }
+
+    /// Decode a Base16 (hex) string into a byte vector.
+    #[cfg(feature = "alloc")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+    pub fn decode_vec(&self, input: impl AsRef<[u8]>) -> Result<Vec<u8>, Error> {
+        let mut output = vec![0u8; Self::decoded_len(input.as_ref())?];
+        self.decode(input, &mut output)?;
+        Ok(output)
+    }
+
+    /// Compute decoded length of the given input.
+    pub fn decoded_len(bytes: &[u8]) -> Result<usize, Error> {
+        if bytes.len() & 1 == 0 {
+            Ok(bytes.len() >> 1)
```

Comment:
It's fine to simply use `bytes.len() / 2`, compiler will replace it with the shift itself. :)